### PR TITLE
Failover changes

### PR DESCRIPTION
**Pull request by tarakreddy1984**

### CODE REVIEW

**Review by tarakreddy1984 — COMMENTED** [review]
updated spell

**Review by tarakreddy1984 — COMMENTED** [review]
updated doc issue

**Review by tarakreddy1984 — COMMENTED** [review]
updated dox

**Review by tarakreddy1984 — COMMENTED** [review]
dox

**Review by tarakreddy1984 — COMMENTED** [review]
dox

**Review by tarakreddy1984 — COMMENTED** [review]
dox

**Review by tarakreddy1984 — COMMENTED** [review]
compile fux

**Review by tarakreddy1984 — COMMENTED** [review]
compile fix

**Review by tarakreddy1984 — COMMENTED** [review]
fix

**Review by tarakreddy1984 — COMMENTED** [review]
compile fix

**Review by kcudnik — APPROVED** [review]

**Comment by lguohan** [comment]
I think we need to have a design doc for this changes, check example here. https://github.com/opencomputeproject/SAI/pull/1122/files#diff-cd05323e7c13f1f9a39ae84ff88f1fab

**Review by lguohan — CHANGES_REQUESTED** [review]
need design doc

**Review by tarakreddy1984 — COMMENTED** [review]
Query for Failove mode support

**Review by tarakreddy1984 — COMMENTED** [review]
updated comment

**Review by tarakreddy1984 — COMMENTED** [review]
changes as per discussion

**Review by tarakreddy1984 — COMMENTED** [review]
updated

File: doc/macsec-gearbox/PAI-Proposal-Failover.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+The PHY failover mode enables a port to broadcast data on both primary and secondary ports. The MAC would be configured to receive the data on either primary or secondary ports. PHY failover active ingress port can be configured as either primary or secondary port. When the link between MAC and PHY goes down on primary port, user can switch to secondary port.
+
+## Failover Configuration on PHY
+
+Failover feature can be configured on all ports of the PHY. Failover ports can operate in either hitless (the ability to switch over port without losing link) or non-hitless mode.
+
+Additions to saiport.h
+
+/\*\*
```
Comment (lguohan) [inline doc/macsec-gearbox/PAI-Proposal-Failover.md:32]: can you use code block to make it clear.

File: doc/macsec-gearbox/PAI-Proposal-Failover.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+\*/
+
+SAI\_SWITCH\_ATTR\_SUPPORTED\_FAILOVER\_MODE,
+
+# Configuration Example
+
+Following example shows how to setup failover configuration on a system side port. In this example a 16 lane PHY is used.
+
+/\* Create System side, Line side, and System Side Failover ports \*/
```
Comment (lguohan) [inline doc/macsec-gearbox/PAI-Proposal-Failover.md:169]: use code block?

**Comment by lguohan** [comment]
can you fix the DCO?

File: doc/macsec-gearbox/PAI-Proposal-Failover.md (hunk 1 above)
Comment (tarakreddy1984) [reply doc/macsec-gearbox/PAI-Proposal-Failover.md:48]: Added block

File: doc/macsec-gearbox/PAI-Proposal-Failover.md (hunk 2 above)
Comment (tarakreddy1984) [reply doc/macsec-gearbox/PAI-Proposal-Failover.md:186]: Configuration Example is in block and added below lines as bold

**Review by tarakreddy1984 — COMMENTED** [review]
Updated doc

File: doc/macsec-gearbox/PAI-Proposal-Failover.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+[1.3 Additions to saiswitch.h](#_Toc52272609)
+
+[2 Configuration Example](#_Toc52272610)
+
+# Overview
+
+The purpose of this document is to describe PHY failover (FO) functionality and interface to manage failover functionality of the PHY. PHY is the connector between MAC (SerDes) and physical medium such as optical fiber or copper transceivers. PHY failover feature uses two physical port configurations: primary and secondary. This can be enabled on system/host or line/client side of a port. It provides the ability to switch from primary port to secondary port or vice versa without losing data/link.
+
+![](RackMultipart20200929-4-1xknbkf_html_6fd49d19699540a6.png)
```
Comment (lguohan) [inline doc/macsec-gearbox/PAI-Proposal-Failover.md:34]: i could not find this file. is it missed?

**Review by tarakreddy1984 — COMMENTED** [review]
added image

**Review by tarakreddy1984 — COMMENTED** [review]
Added image to figures

File: doc/macsec-gearbox/PAI-Proposal-Failover.md (hunk 3 above)
Comment (tarakreddy1984) [reply doc/macsec-gearbox/PAI-Proposal-Failover.md]: Added image to  doc/figures/phy_failover_on_host_side.png

File: doc/macsec-gearbox/PAI-Proposal-Failover.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+[1.3 Additions to saiswitch.h](#_Toc52272609)
+
+[2 Configuration Example](#_Toc52272610)
+
+# Overview
+
+The purpose of this document is to describe PHY failover (FO) functionality and interface to manage failover functionality of the PHY. PHY is the connector between MAC (SerDes) and physical medium such as optical fiber or copper transceivers. PHY failover feature uses two physical port configurations: primary and secondary. This can be enabled on system/host or line/client side of a port. It provides the ability to switch from primary port to secondary port or vice versa without losing data/link.
+
+![](figures/phy_failover_on_host_side.png)
```
Comment (lguohan) [inline doc/macsec-gearbox/PAI-Proposal-Failover.md:32]: the location is not correct. you can put figures in macsec-gearbox directory and reference from file directly using the file name.

**Review by tarakreddy1984 — COMMENTED** [review]
Added correct path for image

**Comment by lguohan** [comment]
for dco, i suggest to squash into one commit and sign-off that commit and then force push.

**Review by lguohan — APPROVED** [review]

**Pull request merged**